### DISCUSSION TITLE
Stop socket.connect after some time

### DISCUSSION THREAD

**Discussion by dst10** [micropython/micropython#11947] [Libraries & Drivers] [answered]
Hi,

I'm trying to program a local network IP scanner in an esp32 board. The purpose of this code is to find which IPs in a local network have port 502 open. For that, I'm using usocket library:

1. Connect esp32 to my local network
2. Creation of the target IPs list: 192.168.1, 192.168.2, 192.168.3,...
3. For each IP, I try to connect to the corresponding port
4. When IP exists and port 502 is open, connection takes place instantly
5. However, when IP doesn't exist or port 502 is not open, socket tries to connect during 30-40 seconds.

Problem: As I want to scan a lot of IPs, to speed up the process I would like to find a way to stop executing _socket.connect_ after 0.5-1 seconds.

Things that I tried:

1. Set _socket.defaulttimeout_ to a very short time. However, this time counts after the program have tried to "resolve" the IP (around 30 seconds)
2. Use __thread_ module and run _socket.connection_ on the second core. However, I didn't find a way to stop the thread and run another one with the next IP

As I understand, my problem isn't a specific problem with socket module. It is more a programming problem for stopping a function after some time if it doesn't end. I can solve it in "normal" Python, but not in Micropython.

If necessary, I can attach the code.

Thanks!

**Comment by karfas** [comment]
It might be a good idea to study the [*fine* documentation](https://docs.micropython.org/en/latest/library/socket.html), set the socket to nonblocking and use `select.poll()` for the timeout.

**Reply by dst10** [reply]
Thanks @karfas, I'll study this documentation in detail because I've never used polling before. I did some testing but I couldn't get anything. After connecting esp32 to my local network, I wrote some code like that:
```
import usocket as socket
import select
s = socket.socket()
poller = select.poll()
poller.register(s.connect(('192.168.1.20',502)), select.POLLIN)
res = poller.poll(1000)
```
However, I can't understand which object must be inside poller.register (I tried both s and s.connect) and also the event flag from the poller

**Comment by karfas** [comment]
Hey, there is an **example** in the linked doc - see the `socket.settimeout()` description.

Something similar should work (can't test this in the moment):
```
s = socket.socket()
s.setblocking(False)
poller = select.poll()
poller.register(s, select.POLLIN)
s.connect(('192.168.1.20',502))
if poller.poll(1000):
    pass 
    # OK, connection succeeds
else
    # No connection within 1 second
    s.close()
```    
Edit: `s.close()` instead of `s.disconnect()`, as there is no disconnect() function.

**Comment by dst10** [comment]
Hi, 

I tried this:

1. Code with `s.setblocking(False)` 
    Returns `OSError: [Errno 119] EINPROGRESS`. For all addresses. No matter if the address really exists or not.

2. Code without `s.setblocking(False)` 
    - For an address that doesn't exist --> Returns `raise PyboardError('timeout waiting for first EOF reception') `
    - For an address that connection is possible, it connects and goes for the `else` path of if loop

The last one, I imagine that it's because s.connect doesn't return anything

**Reply by karfas** [reply]
EINPROGRESS is expected for nonblocking operations.
Just create a try/catch block around the connect().

**Reply by dst10** [reply]
Yes but EINPROGRESS error arises for every address I introduce. And for some of these addresses I know that the connection should be successful.
I've been working on a solution to this problem for a very long time, but I can't solve it

**Reply by karfas** [reply]
I will try this at home in 2-3 hours.
Maybe you need to create the socket as `socket(AF_INET, SOCK_STREAM)`
Did you try to select for select.POLLOUT ?

**Reply by dst10** [reply]
Yes, I tried `select.POLLOUT`, `select.POLLHUP` and `socket(AF_INET, SOCK_STREAM)`. The program behaves as polling isn't implemented.
Thanks for your help @karfas

**Reply by karfas** [reply, marked as answer]
Don't see a problem here.

Tried with
```py
import socket
import select
from time import sleep

def check_port(ip, port):
    s = socket.socket()
    s.setblocking(False)
    poller = select.poll()
    poller.register(s, select.POLLIN | select.POLLOUT)

    try:
        s.connect((ip,port))
    except Exception as e:
        print("Exception connect(): {}".format(e))

    success = ""
    if poller.poll(500):
        success = "OK"
        # OK, connection succeeds
    else:
        success = "NOT OK"
        # No connection within 1 second

    print("connect to {}:{}: {}".format(ip, port, success))
    s.close()

def connect_wlan():
    import network
    wlan = network.WLAN(network.STA_IF)
    wlan.active(False)
    sleep(1)
    wlan.active(True)
    wlan.connect('xxx', 'yyy')
    while not wlan.isconnected():
        print("trying connect...")
        sleep(1)
    print('network config:', wlan.ifconfig())


connect_wlan()
check_port('192.168.8.60', 22) # Linux workstation
check_port('192.168.8.38', 22) # not-existing
check_port('192.168.8.1', 22) # OpenWRT router
```

giving me the (correct) output:
```
network config: ('192.168.8.241', '255.255.255.0', '192.168.8.1', '192.168.8.1')
Exception connect(): [Errno 119] EINPROGRESS
connect to 192.168.8.60:22: OK
Exception connect(): [Errno 119] EINPROGRESS
connect to 192.168.8.38:22: NOT OK
Exception connect(): [Errno 119] EINPROGRESS
connect to 192.168.8.1:22: OK
```

**Reply by karfas** [reply]
Found out this doesn't find closed ports and added a dummy read to account for ECONNRESET.
```
    if poller.poll(500):
        success = "OK"
        r = ""
        try:
            r = s.read(1)
        except Exception as e:
            print("Exception read(): {}".format(e))
            success = "NOT OK"
```

**Comment by dst10** [comment]
Thankss @karfas It works perfectly! I don't know what I was doing wrong when I test what you told me. I was so stuck in this issue. Great help!
